### PR TITLE
Allow comments to only be made when update `patch` matches text

### PR DESCRIPTION
**Pull request by mjgpy3**
**Usage?**

See README and tests.

**Why?**

This should be a simple first step to getting alerted when we're making certain
kinds of changes to our codebase that may require forgotten, missed steps.

**Additional work**

Since this is the "simple first step" :tm:, the structure is intentionally
over-engineered -- trying to make it extendable to future improvements in a
backwards-compatible way.

I'd like to file these enhancements as "good first time" issues, if that's
acceptable to the reviewer, rather than pursuing them all right now. The main
tipping point is that they require more awareness of the PATCH structure.

So, given something like the the README's example, future improvements may be

**Only comment on additions/deletions**, e.g.
```yaml
UnsafeMentionedInCode:
  paths:
    - "backend/**/*.hs"
  where:
    patch:
      additions:
        contain:
          - unsafe
  body: |
    :wave: Hi, I see an addition of "unsafe" in Haskell code. Please consider
    finding a safer alternative!
```

**Comment on regex match**, e.g.
```yaml
UnsafeMentionedInCode:
  paths:
    - "backend/**/*.hs"
  where:
    patch:
      additions_or_deletions:
        match:
          - unsafe[A-Z]
  body: |
    :wave: Hi, I see usage of a "unsafe*" function in Haskell code. If you
    removed it, good going! If you added it, please consider finding a safer
    alternative!
```

**Inline annotation**, e.g.
```yaml
UnsafeMentionedInCode:
  paths:
    - "backend/**/*.hs"
  comment_type: annotation
  where:
    patch:
      additions_or_deletions:
        match:
          - unsafe
  body: |
    :wave: Hi, this looks like a mention of "unsafe" in Haskell code. If you
    removed it, good going! If you added it, please consider finding a safer
    alternative!
```

**Culmination of all these features**, e.g.
```yaml
UnsafeMentionedInCode:
  paths:
    - "backend/**/*.hs"
  comment_type: annotation
  where:
    patch:
      additions:
        match:
          - unsafe[A-Z]
  body: |
    :wave: Hi, this line uses an unsafe Haskell function! Please consider
    finding a safer alternative!
```

### CODE REVIEW

File: src/commenter.ts [hunk 1]
```diff
@@ -97,16 +114,24 @@ async function fetchContent(client: ClientType, path: string): Promise<string> {
...
       }
     }
   }
 
   return false;
 }
+
+const whereIsMet = (patch: string, where: ConfigurationWhereClause) =>
+  where.patch.additions_or_deletions.contain.some((needle) =>
+    patch.includes(needle)
```
Comment (mjgpy3) [inline src/commenter.ts:136]: There's a technical bug here where you could match something in the patch headers (e.g. `@@` or a number). I think that's probably acceptable.

File: src/commenter.ts [hunk 2]
```diff
@@ -68,9 +74,20 @@ async function addComment(client: ClientType, body: string): Promise<void> {
   });
 }
 
+// For now, since there's only 1 type of `where` match we can do, the object is
+// all-or-nothing.
+type ConfigurationWhereClause = {
```
Comment (pbrisbin) [inline src/commenter.ts:79]: Now that this exists, should we consider changing:

```yaml
  paths:
    - "backend/**/*.hs"
  where:
    patch:
      additions_or_deletions:
        contain:
          - unsafe
```

To

```yaml
  where:
    path:
      matches: "backend/**/*.hs"
    patch:
      additions_or_deletions:
        contain:
          - unsafe
```

?

Comment (pbrisbin) [reply src/commenter.ts:79]: For what it's worth, I don't know that I need the `patch` level of this object. "Where additions-or-deletions contain" makes sense to me, it feels fairly obvious which aspect of the PR an "addition" or "deletion" comes from.

**Review by pbrisbin — APPROVED** [review]

**Review by stackptr — APPROVED** [review]

File: src/commenter.ts (hunk 2 above)
Comment (mjgpy3) [reply src/commenter.ts:79]: I like both of these suggestions and will implement them.

**Pull request merged**